### PR TITLE
Mark added E2E testing with ChatServer

### PR DESCRIPTION
**Pull request by markgreenwood**
This took about 3 hrs. Broke out server code into ChatServer.js (except for top level invocation of .listen(). Composed ChatServer end-to-end tests to check:
1. Replies (broadcasts) when a client joins the chat.
2. Broadcasts to the group when additional clients join the chat.
3. Sends client messages to everyone except the sender.
4. Broadcasts to the group when a client leaves the chat.

### CODE REVIEW

**Comment by markgreenwood** [comment]
Added CLI chat client (tcp-chat-client.js). Didn't write any tests for it yet, but it works with manual testing.

File: test/tcp-chat-server.test.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 const assert = require('chai').assert;
-const ChatRoom = require('../ChatRoom');
+const ChatRoom = require('../lib/ChatRoom');
 
-describe('ChatRoom', () => {
+describe('ChatRoom unit tests', () => {
```
Comment (afoldoe) [inline test/tcp-chat-server.test.js]: Thank you for making your it block more semantic!

File: lib/ChatServer.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  client.setEncoding('utf-8');
+  chatRoom.addClient(client);
+  console.log(`${client.name} joined the chat...`);
+  chatRoom.broadcast(`${client.name} joined the chat...`);
+  
+  client.on('data', message => {
+    let re = /\/nick\s+/;
+    if (re.test(message)) {
+      let newName = message.replace(re, '').trim();
+      let bcastNameChange = `${client.name} changing name to ${newName}`;
```
Comment (afoldoe) [inline lib/ChatServer.js]: Liking how well you've taken to template strings!

File: tcp-chat-client.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+const readline = require('readline');
+
+let port = process.argv[3];
+let host = process.argv[2];
+
+const rl = readline.createInterface( { input: process.stdin, output: process.stdout } );
+
+let client = net.connect(port, host, () => {
+
+  console.log(`Connected on port ${port}...`);
```
Comment (afoldoe) [inline tcp-chat-client.js]: Missed commenting out/deleting this line.

File: tcp-chat-client.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  client.setEncoding('utf-8');
+
+  client.on('data', (data) => {
+    console.log(data);
```
Comment (afoldoe) [inline tcp-chat-client.js]: This guy, too.

**Pull request merged**